8347555: [REDO] C2: implement optimization for series of Add of unique value

tabjy opened the pull request:
[JDK-8347555](https://bugs.openjdk.org/browse/JDK-8347555) is a redo of [JDK-8325495](https://bugs.openjdk.org/browse/JDK-8325495) was [first merged](https://git.openjdk.org/jdk/pull/20754) then backed out due to a regression. This patch redos the feature and fixes the bit shift overflow problem. For more information please refer to the previous PR.

When constanlizing multiplications (possibly in forms on `lshifts`), the multiplier is upgraded to long and then later narrowed to int if needed. However, when a `lshift` operand is exactly `32`, overflowing an int, using long has an unexpected result. (i.e., `(1 << 32) = 1` and `(int) (1L << 32) = 0`)

The following was implemented to address this issue.
```java
if (UseNewCode2) {
    *multiplier = bt == T_INT
        ? (jlong) (1 << con->get_int()) // loss of precision is expected for int as it overflows
        : ((jlong) 1) << con->get_int();
} else {
    *multiplier = ((jlong) 1 << con->get_int());
}
```

Two new bitshift overflow tests were added.

dean-long commented:
/reviewers 2

dean-long reviewed (changes requested):
It's not clear what result you are expecting from the new shift code when it overflows.  It looks like signed overflow undefined behavior (UB) to me, which would ideally be caught by UBSAN, so you probably want to make sure your changes are ubsan-clean.  If the desired result for overflow is 0, then I think java_shift_left() should be used.

dean-long commented:
You might want to have the reviewers from the original JDK-8325495 double check this.
@rwestrel @chhagedorn

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -394,9 +394,166 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+        return nullptr;
+      }
+
+      lhs_multiplier = bt == T_INT
```
Can you define: `java_shift_left(jlong, jint, BasicType bt)` that then call one or the other `java_shift_left`?

tabjy commented:
Added `java_shift_left(jlong, jint, BasicType bt)` (and `java_shift_right(...)`, `java_shift_right_unsigned(...)` for the sake of consistency). All tests are passing.

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,165 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+        return nullptr;
+      }
+
+      lhs_multiplier = bt == T_INT
```
Why isn't it: `java_shift_left(1, con->get_int(), bt)` ?

tabjy replied:
You're absolute right. Sorry, that was a very naive oversight.

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,165 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+          return nullptr;
+        }
+
+        *multiplier = lhs_multiplier + (bt == T_INT
```
Same here

eme64 commented on test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:
```diff
@@ -0,0 +1,291 @@
+
+/*
+ * @test
+ * @bug 8325495
```
Should we adjust / add the bug number?

eme64 commented on test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:
```diff
@@ -0,0 +1,291 @@
+        // x = 63 (phi), i = 64 (phi + 1)
+        return i + (x << i) + i; // Expects 64 + 63 + 64 = 191
+    }
+}
```
Would it make sense to add some randomized patterns, just for result verification?

You can use `Generators.java` to get interesting values. Of course that would mean not doing IR verification, but at least it would give us better confidence that the values are correct.

I'm imagining expressions like this:
`return a * CON1 + a * CON2 + a * CON3 + a * CON4`

Where the CON are defined as a `public static final` field with a random value generated by `Generators`.

The advantage of using `Generators` is that it generates powers-of-two more frequently, which seems to be relevant here.

tabjy replied:
That's a very valid suggestion. Thanks. I updated the test. Please take a look if I understood your intention correctly.

However, it does look like this `a * CON1 + a * CON2 + ...` pattern is picked up by existing the associative optimization first:

https://github.com/openjdk/jdk/blob/a23fb0af65f491ef655ba114fcc8032a09a55213/src/hotspot/share/opto/addnode.cpp#L345-L377

eme64 reviewed (changes requested):
This looks interesting! Thanks @tabjy for the work!

For now, I just have some drive-through comments about testing.

Also: would it make sense to have a JMH benchmark to prove that this code change is beneficial enough to warrant the additional complexity?

tabjy commented:
I did some micro benchmark. The result is expected: unoptimized baseline scales linearly as the # of terms, while the optimized remains constant. 

@eme64 Please see [`test/micro/org/openjdk/bench/vm/compiler/SerialAdditions.java`](https://github.com/openjdk/jdk/pull/23506/files#:~:text=test/micro/org/openjdk/bench/vm/compiler/SerialAdditions.java) for details. This is my first time using the JMH framework, please let me know if I made any mistake. Thanks!

---

![](https://github.com/user-attachments/assets/f24c4443-16cc-4efc-9e41-6888f75114a8)

**baseline**: 

```sh
$ CONF=linux-x86_64-server-release make test TEST=micro:org.openjdk.bench.vm.compiler.SerialAdditions
```

```
Benchmark                      Mode  Cnt  Score   Error  Units
SerialAdditions.addIntsMixed   avgt   12  8.231 ± 0.767  ns/op
SerialAdditions.addIntsTo02    avgt   12  0.516 ± 0.074  ns/op
SerialAdditions.addIntsTo04    avgt   12  0.496 ± 0.033  ns/op
SerialAdditions.addIntsTo05    avgt   12  0.492 ± 0.031  ns/op
SerialAdditions.addIntsTo06    avgt   12  0.571 ± 0.029  ns/op
SerialAdditions.addIntsTo08    avgt   12  0.711 ± 0.038  ns/op
SerialAdditions.addIntsTo16    avgt   12  1.245 ± 0.049  ns/op
SerialAdditions.addIntsTo23    avgt   12  1.793 ± 0.054  ns/op
SerialAdditions.addIntsTo32    avgt   12  2.635 ± 0.033  ns/op
SerialAdditions.addIntsTo42    avgt   12  3.726 ± 0.092  ns/op
SerialAdditions.addIntsTo64    avgt   12  6.468 ± 0.215  ns/op
SerialAdditions.addLongsMixed  avgt   12  6.462 ± 0.202  ns/op
SerialAdditions.addLongsTo02   avgt   12  0.467 ± 0.014  ns/op
SerialAdditions.addLongsTo04   avgt   12  0.466 ± 0.022  ns/op
SerialAdditions.addLongsTo05   avgt   12  0.495 ± 0.024  ns/op
SerialAdditions.addLongsTo06   avgt   12  0.553 ± 0.023  ns/op
SerialAdditions.addLongsTo08   avgt   12  0.688 ± 0.032  ns/op
SerialAdditions.addLongsTo16   avgt   12  1.201 ± 0.034  ns/op
SerialAdditions.addLongsTo23   avgt   12  1.736 ± 0.053  ns/op
SerialAdditions.addLongsTo32   avgt   12  2.679 ± 0.088  ns/op
SerialAdditions.addLongsTo42   avgt   12  3.742 ± 0.081  ns/op
SerialAdditions.addLongsTo64   avgt   12  6.359 ± 0.056  ns/op
```

**with this patch**: 

```sh
$ CONF=linux-x86_64-server-release make test TEST=micro:org.openjdk.bench.vm.compiler.SerialAdditions TEST_VM_OPTS="-XX:+UnlockDiagnosticVMOptions -XX:+UseNewCode"
```

```
Benchmark                      Mode  Cnt  Score   Error  Units
SerialAdditions.addIntsMixed   avgt   12  0.933 ± 0.045  ns/op
SerialAdditions.addIntsTo02    avgt   12  0.459 ± 0.015  ns/op
SerialAdditions.addIntsTo04    avgt   12  0.468 ± 0.018  ns/op
SerialAdditions.addIntsTo05    avgt   12  0.467 ± 0.022  ns/op
SerialAdditions.addIntsTo06    avgt   12  0.455 ± 0.020  ns/op
SerialAdditions.addIntsTo08    avgt   12  0.475 ± 0.027  ns/op
SerialAdditions.addIntsTo16    avgt   12  0.470 ± 0.018  ns/op
SerialAdditions.addIntsTo23    avgt   12  0.469 ± 0.028  ns/op
SerialAdditions.addIntsTo32    avgt   12  0.474 ± 0.017  ns/op
SerialAdditions.addIntsTo42    avgt   12  0.476 ± 0.012  ns/op
SerialAdditions.addIntsTo64    avgt   12  0.480 ± 0.017  ns/op
SerialAdditions.addLongsMixed  avgt   12  1.051 ± 0.040  ns/op
SerialAdditions.addLongsTo02   avgt   12  0.460 ± 0.027  ns/op
SerialAdditions.addLongsTo04   avgt   12  0.473 ± 0.028  ns/op
SerialAdditions.addLongsTo05   avgt   12  0.463 ± 0.024  ns/op
SerialAdditions.addLongsTo06   avgt   12  0.457 ± 0.022  ns/op
SerialAdditions.addLongsTo08   avgt   12  0.523 ± 0.074  ns/op
SerialAdditions.addLongsTo16   avgt   12  0.460 ± 0.013  ns/op
SerialAdditions.addLongsTo23   avgt   12  0.468 ± 0.018  ns/op
SerialAdditions.addLongsTo32   avgt   12  0.523 ± 0.078  ns/op
SerialAdditions.addLongsTo42   avgt   12  0.470 ± 0.019  ns/op
SerialAdditions.addLongsTo64   avgt   12  0.475 ± 0.032  ns/op

```
---

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,159 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+// Try to convert a serial of additions into a single multiplication. Also convert `(a * CON) + a` to `(CON + 1) * a` as
+// a side effect. On success, a new MulNode is returned.
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, BasicType bt) {
+  // We need to make sure that the current AddNode is not part of a MulNode that has already been optimized to a
+  // power-of-2 addition (e.g., 3 * a => (a << 2) + a). Without this check, GVN would keep trying to optimize the same
+  // node and can't progress. For example, 3 * a => (a << 2) + a => 3 * a => (a << 2) + a => ...
+  if (find_power_of_two_addition_pattern(this, bt, nullptr) != nullptr) {
```
Where does the optimization `3 * a => (a << 2) + a` happen? Do we use `find_power_of_two_addition_pattern` there too? If not: how do we prevent the code of the two locations from diverging in the future?

tabjy replied:
This `3 * a => (a << 2) + a` happens in `MulINode::Ideal()` and is independent from my code. My code does not explicitly produce power of two patterns but simple multiplication nodes, but such a multiplication node will be later idealized into power of two's. 

Power of two patterns make my code more complex, but this is out of my control.

> how do we prevent the code of the two locations from diverging in the future?

I don't expect primitive opts like power-of-two being removed. Even if it does, `find_simple_multiplication_pattern()` will still work.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,159 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Try to convert a serial of additions into a single multiplication. Also convert `(a * CON) + a` to `(CON + 1) * a` as
```
What about `(a * CON1) + (a * CON2)`? Like `11 * a + 5 * a`. Do we also optimize that?

eme64 replied:
Ah interesting. It could be worth adding a comment for that here then!

eme64 replied:
That was in fact a large part of my initial hesitation with this PR.

tabjy replied:
Added comments on `(a*b) + (a*c)` and `AddNode::IdealIL`

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,159 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+    Node* con = n->in(1);
+    Node* base = n->in(2);
+
+    // swap ConNode to lhs for easier matching
+    if (!con->is_Con()) {
+      swap(con, base);
+    }
```
Is that necessary? Does `Mul` not automatically get canonicalized so that the constant is on the rhs?

tabjy replied:
This is not related to `Mul` canonicalization. Swapping those two variables makes my next 4 lines syntactically easier to write. (So I don't have to do `(con->is_Con ? con : base)->is_top()` and so on.) This shouldn't be any more costly with a modern C++ compiler.

tabjy replied:
Obsolete after refactored to use `find_simple_lshift_pattern`.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,159 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+// given that lhs is different from rhs.
+// Note that one of the term of the addition could simply be `a` (i.e., a << 0). Calling this function with `multiplier`
+// being null is safe.
+Node* AddNode::find_power_of_two_addition_pattern(Node* n, BasicType bt, jlong* multiplier) {
```
This code here looks quite complicated. Why not parse both sides of the add with a `find_simple_lshift_pattern`, and then check that they use the same variable?

tabjy replied:
Good point. Updated to use `find_simple_lshift_pattern`

eme64 commented on test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:
```diff
@@ -0,0 +1,336 @@
+ * @run driver compiler.c2.TestSerialAdditions
+ */
+public class TestSerialAdditions {
+    private static final Random RNG = Utils.getRandomInstance();
```
Could you use `Generators.java` instead? It will produce more "interesting" random values, such as powers of 2, and values close to powers of 2.

eme64 reviewed (commented):
This looks really interesting!

I see that you are doing some special pattern matching. I wonder if it might be worth generalizing the algorithm, to search through an arbitrary "tree" of additions, collect all "leaves" of (`variable * multiplier`), sort by `variable`, and compute new additions for each `variable`. What do you think?

eme64 commented on test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:
```diff
@@ -0,0 +1,336 @@
+    }
+
+    // --- random tests ---
+    private static final int CON1_I, CON2_I, CON3_I, CON4_I;
+    private static final long CON1_L, CON2_L, CON3_L, CON4_L;
+
+    static {
+        RestrictableGenerator<Integer> genI = Generators.G.powerOfTwoInts(16);
+        CON1_I = genI.next();
+        CON2_I = genI.next();
+        CON3_I = genI.next();
+        CON4_I = genI.next();
+
+        RestrictableGenerator<Long> genL = Generators.G.powerOfTwoLongs(16);
+        CON1_L = genL.next();
+        CON2_L = genL.next();
+        CON3_L = genL.next();
+        CON4_L = genL.next();
+    }
```
Is there a reason why you are restricting the values to `powerOfTwoLongs`? I think it would be better if you just take the most general generator.
```
private static final RestrictableGenerator<Integer> GEN_INT = Generators.G.ints();
private static final RestrictableGenerator<Long> GEN_LONG = Generators.G.longs();
```

tabjy replied:
Sorry misunderstood your previous comment. I thought you wanted to test especially power-of-two's. Now I understand you meant more "interesting" randoms (potentially close to power-of-two's)

tabjy commented:
> @eme64  I wonder if it might be worth generalizing the algorithm, to search through an arbitrary "tree" of additions, collect all "leaves" of (`variable * multiplier`) [...]

This was actually my first approach, a top-down approach starts from "root" and collect all leaves to convert to multiplications in one go. It didn't work mainly due to two problems:

1. Not being a technical "tree" introduces additional complexity with repeated nodes (and/or sub-"trees")
    
It's very likely that a subtree contains repeated nodes (e.g., `Add(x, x)`). This is especially problematic if `x` itself is a complex subgraph. Deduplication (probably with memoization) introduces additional complexity. Luckily, even if it's not a tree, it's still strictly a DAG. A bottom-up approach converting leaves incrementally would avoid the issue of duplicated computation.

2. A very large tree results in resource exhaustion on one (i)GVN pass

It's common for some simple loop to be unrolled into a very large subgraph, in which case the transforming the entire graph in one go is blocking and takes a long time to run. By dividing work into individual passes, we allow the optimization to progress and reduce memory footprint.

tabjy commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,159 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  }
+
+  return nullptr;
+}
```
Good suggestion. I didn't make fields `const` as implicitly deleting the constructor makes pattern matching more complex.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,122 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+// Try to convert a serial of additions into a single multiplication. Also convert `(a * CON) + a` to `(CON + 1) * a` as
+// a side effect. On success, a new MulNode is returned.
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, BasicType bt) {
+  // We need to make sure that the current AddNode is not part of a MulNode that has already been optimized to a
+  // power-of-2 addition (e.g., 3 * a => (a << 2) + a). Without this check, GVN would keep trying to optimize the same
+  // node and can't progress. For example, 3 * a => (a << 2) + a => 3 * a => (a << 2) + a => ...
+  if (find_power_of_two_addition_pattern(this, bt).valid) {
+    return nullptr;
+  }
```
Hmm. So somewhere we would have generated that pattern, probably in MulNode. Can you add a verification there, to check that we are only generating patterns that `find_power_of_two_addition_pattern` recognizes? That would make sure that we keep the code here and there in sync.

tabjy replied:
After experiment for a while, I found it not practical to assert power-of-2 patterns in `Ideal()` (or anywhere else).

https://github.com/openjdk/jdk/blob/e9b42dcce268f32bd2ec3c01ac6221073b888538/src/hotspot/share/opto/mulnode.cpp#L263-L266

First, `phase->transform()` could transform `LShiftINode`s into something very different. Second, there is no guarantee the `res = AddINode`  will remain untransformed by the time it's passed into `find_power_of_two_addition_pattern()`. Asserting power-of-2 patterns before such transformation is not very helpful.

> @eme64: [...] check that we are only generating patterns that find_power_of_two_addition_pattern recognizes

In short, we can't guarantee we'll always pick up *transformed* power-of-2 patterns, at least not with significant effort and difficulty predicting all possible transformations. However, I would argue making this guarantee is outside the scope of this issue as all serial addition patterns are correctly picked up and optimized right now.

I'm not sure how to proceed with this. Please let me know what you think. Thanks!

eme64 replied:
Hmm, ok I see. Why don't you remove the asserts for now, and we see how clear the code looks now. I think I asked for the consistency check because I was confused by the previous code structure. Maybe it is ok now as it is.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,122 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+
+  // While multiplications can be potentially optimized to power-of-2 subtractions (e.g., a * 7 => (a << 3) - a),
+  // (x - y) + y => x is already handled by the Identity() methods. So, we don't need to check for that pattern here.
+  Multiplication mul;
+  if (
+      ((mul = find_simple_addition_pattern(in1, bt)).valid && mul.variable == in2) ||
+      ((mul = find_simple_lshift_pattern(in1, bt)).valid && mul.variable == in2) ||
+      ((mul = find_simple_multiplication_pattern(in1, bt)).valid && mul.variable == in2) ||
+      ((mul = find_power_of_two_addition_pattern(in1, bt)).valid && mul.variable == in2)
+      ) {
```
I find this quite difficult to read. And it looks repetitive too. Maybe you can refactor it?
Also, it would be nice to have comments with the patterns here, to see which one covers what case, so that we have a nice overview.

tabjy replied:
Refactored. Added comments with example. It's still some-what repetitive and definitely much clearer.

https://github.com/openjdk/jdk/blob/eefe3a35effc469f3527d4825c9cc4f2a72742dc/src/hotspot/share/opto/addnode.cpp#L421-L444

eme64 commented:
You match cases like `(a * CON) + a`.
- What do we do with patterns `a + (a * CON)`?
- What about `(a << CON1) + (a * CON2)` and other generalizations that have mixed operations on either side? You mentioned that things like `(a*b) + (a*c)` -> `a*(b + c)` are already handled by `AddNode::IdealIL`.

It would just be a shame to have all the complexity of matching specific cases, but not take the chance to make it a bit more general.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,122 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+      ((mul = find_power_of_two_addition_pattern(in1, bt)).valid && mul.variable == in2)
+      ) {
+    Node* con = (bt == T_INT)
+                ? (Node*) phase->intcon((jint) (mul.multiplier + 1)) // intentional type narrowing to allow overflow at max_jint
+                : (Node*) phase->longcon((mul.multiplier + 1));
```
I think just to be safe, you should use `java_add` to have correct overflow semantics. You are using `jlong` for `multiplier`, which is a signed integer type, and in C++ overflow is undefined behavior as far as I know, let's avoid that ;)

Actually, do you have a test where the multiplier overflows here?

tabjy replied:
Good point regarding UB. Updated to use `java_add`.

There are several test cases named `.*Overflow()` in `TestSerialAdditions`.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,122 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+      return Multiplication{true, lhs.variable, lhs.multiplier + rhs.multiplier};
+    }
+
+    if (lhs.valid && lhs.variable == n->in(2)) {
+      return Multiplication{true, lhs.variable, lhs.multiplier + 1};
+    }
+
+    if (rhs.valid && rhs.variable == n->in(1)) {
+      return Multiplication{true, rhs.variable, rhs.multiplier + 1};
+    }
```
Hmm, it seems these are patterns that you did not promise you would cover in the description above.
It makes it a little difficult to keep the overview...

chhagedorn replied:
Just a drive-by comment what might help: Name the cases you cover in the description with `(1)`, (2)` etc. and add the numbers as comments in the code where you cover the patterns. This would support the mapping from description to implementation.

tabjy replied:
I updated and expanded the comment. They are clearer now. I also added a subcase for `a + a` so `assert()` in `MulNode::Ideal()` can call directly into `find_power_of_two_addition_pattern()`

> Name the cases you cover in the description [..]

That's a great suggestion! Thank you. I'll start using this technique from now on.

eme64 commented:
@tabjy One more comment: I have had bad experiences before with pattern matching that only covered a part of the cases, and where methods did sometimes do more than what they promised in their name or documentation. These things tend to get extended later, and the overview gets worse and worse until nobody has the overview and bugs creep in that are hard to discover in a review.

Can you do some experimenting and see if you can come up with a cleaner design? Maybe write down at the beginning in `convert_serial_additions` what is the general form of the patterns you cover?

dean-long commented on src/hotspot/share/utilities/globalDefinitions.hpp:
```diff
@@ -1260,6 +1260,24 @@ JAVA_INTEGER_SHIFT_OP(>>, java_shift_right_unsigned, jlong, julong)
+JAVA_INTEGER_SHIFT_BASIC_TYPE(java_shift_left)
+JAVA_INTEGER_SHIFT_BASIC_TYPE(java_shift_right)
+JAVA_INTEGER_SHIFT_BASIC_TYPE(java_shift_right_unsigned)
+
```
Couldn't this be written as a template, so we could use java_shift_left<T_INT>(x,y) or java_shift_left<jint>(x.y)?

tabjy replied:
Explicit `BasicType` as a function (instead of template) parameter allows more flexibility at runtime when `bt` is a variable. For example:

https://github.com/openjdk/jdk/blob/e9b42dcce268f32bd2ec3c01ac6221073b888538/src/hotspot/share/opto/addnode.cpp#L471

eme64 commented:
@tabjy Do you want us to review again? We generally wait for a ping, otherwise we assume you are still working on it ;)

eme64 commented:
@tabjy Thanks for your patience, this one took me longer than I wanted. I responded like this above:

> Hmm, ok I see. Why don't you remove the asserts for now, and we see how clear the code looks now. I think I asked for the consistency check because I was confused by the previous code structure. Maybe it is ok now as it is.

tabjy commented:
Sorry I forgot to ping you @eme64 this time. I addressed all change requests as best as I could and I think another review would be appropriate. Thanks!

eme64 commented on src/hotspot/share/opto/addnode.hpp:
```diff
@@ -42,7 +42,24 @@ typedef const Pair<Node*, jint> ConstAddOperands;
   virtual uint hash() const;
-public:
+
+  struct Multiplication {
+    bool valid = false;
+    Node* variable = nullptr;
+    jlong multiplier = 0;
+  };
+
+  inline static bool is_valid_multiplication(const Multiplication& mul, const Node* variable) {
+    return mul.valid && mul.variable == variable;
+  }
```
Why not make it a `is_valid_with(variable)` method of the struct?

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,164 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+// transformed power-of-2 patterns when term `a` is complex.
+//
+// Therefore, when matching the LHS `a * CON`, we match with best efforts by looking for the following patterns:
+//     - (1) Simple addition: a + a
+//     - (2) Simple lshift: a << CON
+//     - (3) Simple multiplication: CON * a
+//     - (4) Power-of-two addition: (a << CON1) + (a << CON2)
```
```suggestion
//     - (1) Simple addition:       LHS = a + a
//     - (2) Simple lshift:         LHS = a << CON
//     - (3) Simple multiplication: LHS = CON * a
//     - (4) Power-of-two addition: LHS = (a << CON1) + (a << CON2)
```
I suggest adding the `LHS =` explicitly here. I tend to not always read all the explanatory text, and search for patterns first. Then it is not immediately clear that you are actually matching a `a + a + a` rather than a `a + a` for `1)`.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,164 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+//     => 3*a + ... + a
+//     => n*a
+//
+// Due to the iterative nature of iGVN, MulNode transformed from first few AddNode terms may be further transformed into
```
```suggestion
// Due to the iterative nature of iGVN, MulNode transformed from first few AddNode terms may be further transformed into
```
```suggestion
// Due to the iterative nature of IGVN, MulNode transformed from first few AddNode terms may be further transformed into
```

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,164 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+    return nullptr;
+  }
+
+  Node* in1 = in(1);
+  Node* in2 = in(2);
```
For consistency, you should either only use `LHS` / `RHS`, or only `in1` / `in2`.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,164 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  Node* in1 = in(1);
+  Node* in2 = in(2);
+
+  // (1) Simple addition pattern (e.g., a + a)
```
```suggestion
  // (1) Simple addition pattern (e.g., in1 = a + a)
```
Do the same below.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,164 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  return Multiplication{};
+}
+
+// Try to match `a << CON`. On success, return a struct with `.valid = true`, `variable = a`, and
```
```suggestion
// Try to match `n = a << CON`. On success, return a struct with `.valid = true`, `variable = a`, and
```
Do the same below.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,164 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+    }
+  }
+
+  return Multiplication{};
```
```suggestion
  return Multiplication::make_invalid();
```
I would prefer wrapping this in something that is immediately clear at the call site, so the reader does not have to go look at the internals of `Multiplication`.

eme64 reviewed (changes requested):
Did a quick pass over half of the code before finishing for the day.

I still think that the patterns covered are more limited than preferable. It would be nice if we could cover more cases, especially more cases where the RHS has different patterns too.

One question: what if LHS and RHS are flipped, can we recognize patterns like this: `a + (a << CON)`?

tabjy commented:
@eme64 
> I still think that the patterns covered are more limited than preferable. [...] can we recognize patterns like this: a + (a << CON)?

I only aimed to cover cases minimally needed for `a + a + ... + a` at fear of adding even more complexity. Cases like `CON * a + a` are considered unintended side-effects due to the way pattern matching is implemented. Flipped LHS and RHS like `a + (a << CON)` is not recognized (by me or `IdealIL()`). 

That is, I only match the RHS being a base variable (i.e., `a`).  I understand you think giving the effort, we could explore more opportunities here. To explain the complexity, consider the following expression with base variable `a`:

```
   a + a + ((a + a) + a)
     ^ LHS(a) = RHS(a)
= 2*a + ((a + a) + a)
      ^ unable to resolve RHS to 3*a without some recursion
```

I could, at very least, try to swap LHS and RHS if no match is found, but the case above will still not benefit from swapping (without a recursive algorithm). Simpler cases like `a + (a << CON)` might.

tabjy commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -395,9 +395,164 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+//     - (4) Power-of-two addition: (a << CON1) + (a << CON2)
+//
+// Note this also converts, for example, original expression `(a*3) + a` into `4*a` and `(a<<2) + a` into `5*a`. A more
+// generalized pattern `(a*b) + (a*c)` into `a*(b + c)` is handled by AddNode::IdealIL().
```
Please see https://github.com/openjdk/jdk/pull/23506#issuecomment-2851911494

tabjy commented:
Ping @eme64 again for awareness. :)

tabjy commented:
/open

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,179 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  Node* lhs = in(1);
+  Node* rhs = in(2);
+
+  Multiplication mul = find_serial_addition_patterns(lhs, rhs, bt);
+  if (!mul.is_valid_with(rhs)) {
+    // Swap lhs and rhs then try again
+    mul = find_serial_addition_patterns(rhs, lhs, bt);
+    if (!mul.is_valid_with(lhs)) {
+      return nullptr;
+    }
+  }
```
Nice, thanks for adding it!
I think it would be nice if we renamed `find_serial_addition_patterns` so that it is clear that we are looking for `a + a * con` or `con*a + a`. Because currently it is not directly clear why we need the swapping from the method name.

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,179 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  return MulNode::make(con, mul.variable, bt);
+}
+
+// Find a pattern of serial additions that can be converted to a multiplication.
+// When matching the LHS `a * CON`, we match with best efforts by looking for the following patterns:
+//     - (1) Simple addition:       LHS = a + a
+//     - (2) Simple lshift:         LHS = a << CON
+//     - (3) Simple multiplication: LHS = CON * a
+//     - (4) Power-of-two addition: LHS = (a << CON1) + (a << CON2)
+AddNode::Multiplication AddNode::find_serial_addition_patterns(const Node* lhs, const Node* rhs, BasicType bt) {
```
Here, we have `rhs = a`, right? I'd suggest just renaming the method arguments `rhs`->`a` and `lhs`->`pattern`. Because you already call (1) - (4) patterns in the documentation. That would be a good fit :)

eme64 commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,179 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+// Try to match `n = (a << CON1) + (a << CON2)`. On success, return a struct with `.valid = true`, `variable = a`, and
+// `multiplier = (1 << CON1) + (1 << CON2)`.
+// Match `n` for patterns:
+//     - (1) AddNode(LShiftNode(a, CON), LShiftNode(a, CON))
+//     - (2) AddNode(LShiftNode(a, CON), a)
+//     - (3) AddNode(a, LShiftNode(a, CON))
+//     - (4) AddNode(a, a)
```
You could drop the `Node` part from the cases here, to make it a bit more concise.

eme64 replied:
Alternatively, you could do it with the `<<` operator like you did in `find_serial_addition_patterns`. I think that would be more consistent.

eme64 commented on test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:
```diff
@@ -0,0 +1,430 @@
+/*
+ * @test
+ * @bug 8325495 8347555
+ * @summary C2 should optimize for series of Add of unique value. e.g., a + a + ... + a => a*n
```
You may want to change the summary here, and also the PR summary. Because you really do not just do these series of additions, but lots of other cases as well. The examples below suggest that too ;)

eme64 commented on test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:
```diff
@@ -0,0 +1,430 @@
+ * questions.
+ */
+
+package compiler.c2;
```
I would put the test in a more specific directory. I think the `igvn` directory would be a good canditate, because `Ideal` is part of IGVN ;)

eme64 commented on test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:
```diff
@@ -0,0 +1,430 @@
+        }
+    }
+
+    // We can't do IR verification but only check for correctness for a better confidence.
+    @Test
+    private static int randomPowerOfTwoAddition(int a) {
+        return a * CON1_I + a * CON2_I + a * CON3_I + a * CON4_I;
+    }
+
+    @Test
+    private static long randomPowerOfTwoAdditionL(long a) {
+        return a * CON1_L + a * CON2_L + a * CON3_L + a * CON4_L;
+    }
```
Nice, thanks for these :)

eme64 reviewed (changes requested):
@tabjy Thanks for the ping. Sorry I did not respond earlier. I was hoping others would continue the review, but it seems it got stuck on me here, a classic though unfortunate pattern ;)

@rwestrel Asked me if I wanted to continue reviewing. I'm going on a 3-week vacation, so feel free to ask others to review.

--------------------------------

I'll summarize my thoughs now, so others can review the PR in my absence:
- The PR looks much better now, we have made good progress.
- I'm still sad that we are not covering cases like `a * CON1 + a * CON2`, or other patterns that could be collapsed to `a * CON`. But I do understand that this would require some recursive approach, and that could be a little more difficult.

-----------------------------------

I'll leave it at this, and hope that others will review :blush:

tabjy commented:
@eme64 Thank you for reviewing! Those are very valid suggestion, especially on naming as this PR evolves. I've done the following:

- updated naming (mostly with "serial addition" to "collapsable addition (into multiplication)")
- updated comments
- moved test file

Please enjoy your time off!

Once GHA passes, @rwestrel could you please give this a quick review if you have some time? Thank you very much!

tabjy commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,179 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+//
+// Note this also converts, for example, original expression `(a*3) + a` into `4*a` and `(a<<2) + a` into `5*a`. A more
+// generalized pattern `(a*b) + (a*c)` into `a*(b + c)` is handled by AddNode::IdealIL().
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, BasicType bt) {
```
Thank you. I also like the wording of *collapsing* additions better. I've updated names and comments accordingly.

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,182 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+    Multiplication rhs = find_simple_lshift_pattern(n->in(2), bt);
+
+    // Pattern (1)
+    if (lhs.valid && rhs.valid && lhs.variable == rhs.variable) {
```
Would it make sense to add a method to `Multiplication`, let's say `add`. The 3 `if`s here would then be replaced by:
```
Multiplication res = lhs.add(rhs);
if (res.valid()) {
  return res;
}
return find_simple_addition_pattern(n, bt);
```
and the logic from the 3 ifs would be moved to `Multiplication::add`.
What do you think?

tabjy replied:
I added `add(Mulitplication)` and moved `Pattern (1)` to it. However, for pattern `(2)` and `(3)`, the one side is not a `Multiplcation` but a simple node (`(a << CON) + a` and `a + (a << CON)`). Therefore I left the two other `if`s here.

tabjy commented on src/hotspot/share/opto/addnode.hpp:
```diff
@@ -42,7 +42,28 @@ typedef const Pair<Node*, jint> ConstAddOperands;
+    }
+  };
+
+  static Multiplication find_collapsible_addition_patterns(const Node* a, const Node* pattern, BasicType bt);
```
Move to static members of `Multiplication`

rwestrel commented on src/hotspot/share/opto/addnode.hpp:
```diff
@@ -42,7 +42,51 @@ typedef const Pair<Node*, jint> ConstAddOperands;
+
+    Multiplication add(const Multiplication rhs) const {
+      if (is_valid_with(rhs.variable()) && rhs.is_valid_with(variable())) {
+        return {variable(), java_add(multiplier(), rhs.multiplier())};
```
You should use a constructor call here

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,185 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+
+    // Pattern (2)
+    if (lhs.is_valid_with(n->in(2))) {
+      return {lhs.variable(), java_add(lhs.multiplier(), static_cast<jlong>(1))};
```
You should use a constructor call here.

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,185 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+
+    // Pattern (3)
+    if (rhs.is_valid_with(n->in(1))) {
+      return {rhs.variable(), java_add(rhs.multiplier(), static_cast<jlong>(1))};
```
You should use a constructor call here.

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,185 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+// The method matches `n` for pattern: a + a.
+AddNode::Multiplication AddNode::Multiplication::find_simple_addition_pattern(const Node* n, BasicType bt) {
+  if (n->Opcode() == Op_Add(bt) && n->in(1) == n->in(2)) {
+    return {n->in(1), 2};
```
You should use a constructor call here.

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,185 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (n->Opcode() == Op_LShift(bt) && n->in(2)->is_Con()) {
+    Node* con = n->in(2);
+    if (!con->is_top()) {
+      return {n->in(1), java_shift_left(1, con->get_int(), bt)};
```
You should use a constructor call here.

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,185 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+    }
+
+    if (!con->is_top()) {
+      return {base, con->get_integer_as_long(bt)};
```
You should use a constructor call here.

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,185 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  // This optimization technically only produces MulNode(CON, a), but we might as match MulNode(a, CON), too.
+  if (n->Opcode() == Op_Mul(bt) && (n->in(1)->is_Con() || n->in(2)->is_Con())) {
+    // Pattern (1)
+    Node* con = n->in(1);
```
Isn't con always input 2 because `MulNode::Ideal` canonicalize it?

rwestrel commented on src/hotspot/share/opto/addnode.cpp:
```diff
@@ -396,9 +396,185 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+    }
+  }
+
+  Node* con = (bt == T_INT)
```
With an `if`, the `static_cast<Node*>` are not needed, right? I would do that then. It would more readable.
